src/server/journal/journal_slice.cc

```diff
@@ -191,20 +191,26 @@ void JournalSlice::AddLogRecord(const Entry& entry, bool await) {
     DVLOG(2) << "AddLogRecord: run callbacks for " << entry.ToString()
              << " num callbacks: " << change_cb_arr_.size();
 
+    auto callbacks_num = change_cb_arr_.size();
```

Comment:
I cant be sure that this solution is not with bugs
If we have some place where preempt between callback execution and the call to AddLogRecord this solution still has the bug.

What we want is to make the register to dbsilce and journal atomic, so what we can do is lock before transaction callback is executed (Transaction::RunCallback) and release the lock after we write the journal change LogAutoJournalOnShard

Reply:
You FiberAtomicGuard?